### PR TITLE
Refactoring Go Code

### LINKED ISSUE #5: Automated messages should have a better format
Reduce the automated message size.
Possibly move most of the output into the log file.

Move all the outputs to the #all_hooks  channel on slack.

Give direct URL to open the deployed app in preview message.

### LINKED ISSUE #16: DeployBot is not Thread Safe
There are a bunch of operations which mutate the state of:

1. Internal counter variables
2. History files on the disk

These operations are not thread safe, and running them parallely on two diffferent requests can lead to inconsistent state. We need to:

1. Use atomic for small in-memory variables
2. Need to think of a better way to store the history (A periodic go routine to dump the memory state and use mutexes to modify). I feel there is already a golang library to do so.

### LINKED ISSUE #17: #CodeHealth Break into packages
The Deploybot is now in a stage that a single package gets too complicated to understand and there are a lot of exported functions/types which are unncessary.

We need to split it into multiple packages, and export the functions and types used by other packages.

### LINKED ISSUE #21: Implement healtchecking library
I have modified the history file's structs, to store a healtcheck as well.

We need to add a controller that periodically gets services and healthchecks them and stores the state.

### LINKED ISSUE #22: Log the user
We need to log the user, that initiated the action both from slack and git.

Slack should probably give us the userID and we can at initialization store a mapping of the username and userid.

### PR DESCRIPTION
**Pull request by aagrxyz**
This is a starting commit in a chain of commits to refactor the whole DeployBot go code. There are a bunch of issues related to design and
style, and this needs to be fixed, if we want to keep this project going
and add functionalities to it. Couple of things to note:

1. Never panic or log.Fatal from any function other than main
2. We need to pass errors up to the caller and let the caller handle the
error and decide if to panic or not
3. This code is getting bigger, and we need to split it into packages
4. We should try really hard to not modify global variables in
functions. This makes the flow of control completely hard to understand.
If required, we should only update them in functions very close to main
and/or the init() function

This commit -- Refactor the slack.go, helper.go and deploy.go

DO NOT SUBMIT right now.

### CODE REVIEW

File: src/deploy.go [hunk 1]
```diff
@@ -50,79 +41,67 @@ func deployGoRoutine(callbackID string,
...
 	branch := DefaultBranch
 
-	status := GetStatus(gitRepoURL)
+	status, err := GetStatus(gitRepoURL)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get current status of service(%s) - %v", gitRepoURL, err)
+	}
 
 	var output []byte
-	var err error
```
Comment (imperfect-fourth) [inline src/deploy.go:64]: "err" on line 102 will be out of scope

File: src/deploy.go [hunk 2]
```diff
@@ -3,35 +3,27 @@ package main
...
-	submissionDataMap map[string]interface{}) {
-	if chatPostMessage(submissionDataMap["channel"].(string),
-		"Deployment in Progress", nil) == false {
-		log.Warn("Some error occured")
+func deployGoRoutine(callbackID string, submissionDataMap map[string]interface{}) {
+	if err := chatPostMessage(submissionDataMap["channel"].(string), "Deployment in Progress", nil); err != nil {
+		log.Errorf("cannot post begin deployment chat message - %v", err)
 		return
 	}
+	log.Infof("beginning deployment of %s on server %s with subdomain as %s with callback_id as %s", submissionDataMap["git_repo"], submissionDataMap["server_name"], submissionDataMap["subdomain"], callbackID)
```
Comment (imperfect-fourth) [inline src/deploy.go]: line too long. why not wrap it?

Comment (aagrxyz) [reply src/deploy.go]: I had been using gofmt to format all the code, I write. That is the standard way to format go code. 
The official doc says - https://golang.org/doc/effective_go.html#formatting 
`Go has no line length limit. Don't worry about overflowing a punched card. If a line feels too long, wrap it and indent with an extra tab.`
I will not stick to a hard 80 char limit, but format lines, which are too big later.

File: src/deploy.go (hunk 1 above)
Comment (aagrxyz) [reply src/deploy.go:64]: Nope, it is defined on line 52

**Comment by aagrxyz** [comment]
I have started breaking it into packages. Expect more commits on this PR in next 2-3 days.

**Comment by aagrxyz** [comment]
More commits to correct bugs in the history package. @Divi09112 Can you check this version out, build (it builds, I verified that) and run on the server and see if it crashes anywhere. We are using dynamic type reflection a lot and they can cause some issues at runtime.

**Comment by aagrxyz** [comment]
I think this, is safe to merge now. Merging.

**Pull request merged**